### PR TITLE
Adds guard statements to Erlang and Elixir factorial

### PR DESCRIPTION
**Pull request by FrankKair**
**Fixes issue:**
No particular issue.

**Changes:**
Adds `guard` statements to Erlang and Elixir implementations of factorial.

### CODE REVIEW

File: code/mathematical_algorithms/src/factorial/factorial.erl [hunk 1]
```diff
@@ -1,5 +1,5 @@
 -module(factorial).
--export([fac/1]).
+-export([factorial/1]).
 
-fac(0) -> 1;
-fac(N) -> N * fac(N-1).
\ No newline at end of file
+factorial(0) -> 1;
```
Comment (arnavb) [inline code/mathematical_algorithms/src/factorial/factorial.erl]: Is there any way to just return 1 for a negative initial argument?

File: code/mathematical_algorithms/src/factorial/factorial.ex [hunk 2]
```diff
@@ -1,7 +1,7 @@
 defmodule Factorial do 
   def factorial(0), do: 1
```
Comment (arnavb) [inline code/mathematical_algorithms/src/factorial/factorial.ex]: Same here

File: code/mathematical_algorithms/src/factorial/factorial.erl (hunk 1 above)
Comment (FrankKair) [reply code/mathematical_algorithms/src/factorial/factorial.erl]: That would be 

```erlang
factorial(N) when N < 0 -> 1.
```

Comment (FrankKair) [reply code/mathematical_algorithms/src/factorial/factorial.erl]: Do you think it should be handled like so?
Should the answer be yes, I can refactor the code to include the advent of negative numbers.

File: code/mathematical_algorithms/src/factorial/factorial.ex (hunk 2 above)
Comment (FrankKair) [reply code/mathematical_algorithms/src/factorial/factorial.ex]: Yep

```elixir
def factorial(n) when n < 0, do: 1
```

File: code/mathematical_algorithms/src/factorial/factorial.erl (hunk 1 above)
Comment (arnavb) [reply code/mathematical_algorithms/src/factorial/factorial.erl]: Yes, it should. You should probably handle all cases.

File: code/mathematical_algorithms/src/factorial/factorial.ex (hunk 2 above)
Comment (arnavb) [reply code/mathematical_algorithms/src/factorial/factorial.ex]: Yes, it should. You should probably handle all cases.

File: code/mathematical_algorithms/src/factorial/factorial.erl (hunk 1 above)
Comment (OvermindDL1) [reply code/mathematical_algorithms/src/factorial/factorial.erl]: Shouldn't `<0` cause a matcherror since factorial itself is undefined when `<0`?

Comment (arnavb) [reply code/mathematical_algorithms/src/factorial/factorial.erl]: That works as well. I just want it to be handled.

Comment (OvermindDL1) [reply code/mathematical_algorithms/src/factorial/factorial.erl]: Well a matcherror is thrown/raised when there is no matching function head, so technically it is already handled.  :-)

Comment (FrankKair) [reply code/mathematical_algorithms/src/factorial/factorial.erl]: (This also applies to the **Elixir** implementation).

Currently:

```erlang
factorial(0) -> 1;
factorial(N) when N > 0 -> N * factorial(N - 1).
```

The way is defined right now, if you call `factorial(-1)`, there will be no matching function in this Erlang module, so an error will be thrown.

If we define:

```erlang
factorial(N) when N < 0 -> 1.
```

We'll be returning 1 for every negative number (which... is not the definition of factorial).

Comment (FrankKair) [reply code/mathematical_algorithms/src/factorial/factorial.erl]: It's up to you all, but at least for me, it makes more sense to keep the mathematical definition of factorial, not allowing the user to call factorial of negative numbers.

Comment (arnavb) [reply code/mathematical_algorithms/src/factorial/factorial.erl]: Ok, what I thought would happen was that if you called the function with a negative number, then the program would just not work. (Infinite loop). However, it seems that Elixir handles this very well! (Sorry for my incompetence; I don't know Elixir very well, it's just that I am going by general programming paradigms that _all_ cases should be handled).

Comment (FrankKair) [reply code/mathematical_algorithms/src/factorial/factorial.erl]: Hey, @arnavb, no problem. Erlang and Elixir are not very common programming languages and they have some different concepts!

Comment (OvermindDL1) [reply code/mathematical_algorithms/src/factorial/factorial.erl]: Heh yep, they are fully immutable functional languages that also have exceptions, and running through a set of matchers with no match ends up raising a `MatchError`.  :-)

They are languages designed to get real work done, so they have a lot of ways of doing things and are in general robust.  :-)

**Comment by FrankKair** [comment]
@arnavb Is this ready to go, then?

**Comment by arnavb** [comment]
@FrankKair Yup, just wait until @AdiChat merges this!

**Comment by AdiChat** [comment]
Looks great 😍 

Hope you are enjoying your journey 🚆  with **OpenGenus**

**Pull request merged**